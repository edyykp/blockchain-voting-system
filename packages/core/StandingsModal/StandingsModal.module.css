.container {
  display: block;
  font-family: Cambria, Cochin, Georgia, Times, 'Times New Roman', serif;
}
.driverWrapper {
  display: grid;
  grid-template-columns: 70px 1fr 150px 1fr 80px;
  width: 100%;
  height: 80px;
  border-bottom: 0.5px solid var(--gray-400);
  padding-left: 10px;
  align-items: center;
  transition: 0.2s all ease-in-out;
  font-size: 1.1rem;
}

.greyBackground {
  background-color: var(--gray-200);
}
.nameWrapper {
  letter-spacing: 0.3mm;
  display: grid;
  grid-template-columns: 25px auto;
}

.positionWrapper {
  font-weight: 600;
}

.permanentNumber {
  font-family: 'Montagu Slab', serif;
  font-size: 0.9rem;
  letter-spacing: normal;
  color: var(--gray-dark);
}

.constructorWrapper {
  vertical-align: top;
  margin-left: 5px;
}

.arrow {
  font-weight: bold;
  letter-spacing: -0.5mm;
  font-size: 0.9rem;
}

.greenArrow {
  color: var(--green-success);
  text-shadow: 0px 0px 12px var(--green-success);
}

.redArrow {
  color: var(--primary);
  text-shadow: 0px 0px 12px var(--primary);
}

@media screen and (max-width: 650px) {
  .driverWrapper {
    grid-template-columns: auto 60%;
    font-size: 1rem;
  }

  .permanentNumber {
    font-size: 0.8rem;
  }

  .votesWrapper,
  .teamWrapper {
    height: 60px;
    padding-top: 10px;
  }

  .nationalityWrapper {
    display: none;
  }
}
