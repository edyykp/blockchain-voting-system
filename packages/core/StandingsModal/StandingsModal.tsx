import { useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import Image from 'next/image';

import { Modal } from '@packages/components';
import { DriverType } from '@packages/types';

import styles from './StandingsModal.module.css';

export type StandingsModalProps = {
  show: boolean;
  setShowModal: (show: boolean) => void;
  raceName?: string;
  circuitId?: string;
};

export const StandingsModal = ({
  show,
  raceName,
  circuitId,
  setShowModal,
}: StandingsModalProps) => {
  const [drivers, setDrivers] = useState<DriverType[]>([]);
  const [error, setError] = useState<string | null>(null);
  const { query } = useRouter();
  const currentYear = new Date().getFullYear();
  const renderedYear = query.year ?? currentYear;

  const getDrivers = async () => {
    const url = circuitId
      ? `/api/getDrivers?year=${renderedYear}&circuit=${circuitId}`
      : `/api/getDrivers?year=${renderedYear}`;
    const data = await fetch(url);
    const text: { drivers: DriverType[]; error: string | null } =
      await data.json();

    setDrivers(text.drivers);
    setError(text.error);
  };

  useEffect(() => {
    if (show) {
      getDrivers();
    }

    if (!show) {
      setError(null);
    }
  }, [show]);

  useEffect(() => {
    if (error) {
      setShowModal(false);
    }
  }, [error]);

  const positionMovement = (
    startingPosition: number,
    finalPosition: number,
  ) => {
    if (startingPosition === 0) {
      return undefined;
    }
    const difference = Math.abs(
      Number(startingPosition) - Number(finalPosition),
    );

    if (difference === 0) {
      return undefined;
    }

    if (startingPosition > finalPosition) {
      return '▲' + String(difference);
    }

    return '▼' + String(difference);
  };

  const setImageSource = (source: string) => {
    try {
      return require(`public/${source}.jpg`);
    } catch (error) {
      return require('public/placeholder.png');
    }
  };

  const DriversList = (driversList: DriverType[]) => (
    <div className={styles.container} data-testid="standings-container">
      {driversList.map((driver, key) => (
        <div
          className={`${styles.driverWrapper} ${
            key % 2 === 1 ? styles.greyBackground : ''
          }`}
          key={key}
          data-testid="driver-row"
        >
          <div className={styles.positionWrapper}>{key + 1}.</div>
          <div className={styles.nameWrapper}>
            <span className={styles.permanentNumber}>
              {driver.Driver.permanentNumber}
            </span>{' '}
            {driver.Driver.givenName} {driver.Driver.familyName}
          </div>
          <div className={styles.nationalityWrapper}>
            {driver.Driver.nationality}
          </div>
          <div className={styles.teamWrapper}>
            <span className={styles.imageWrapper}>
              <Image
                src={setImageSource(driver.Constructor.constructorId)}
                loading="lazy"
                alt={driver.Constructor.constructorId}
                width={15}
                height={20}
                className={styles.img}
              />
            </span>

            <span className={styles.constructorWrapper}>
              {driver.Constructor.name}
            </span>
          </div>
        </div>
      ))}
    </div>
  );

  return (
    <Modal
      show={show}
      onClose={() => setShowModal(false)}
      children={DriversList(drivers)}
      title={`Standings for ${raceName ?? renderedYear}`}
      theme="light"
    />
  );
};
