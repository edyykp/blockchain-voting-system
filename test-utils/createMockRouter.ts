import { NextRouter } from 'next/router';

export function createMockRouter(router: Partial<NextRouter>): NextRouter {
  return {
    basePath: '',
    pathname: '/',
    route: '/',
    asPath: '/',
    query: {},
    isFallback: false,
    back: jest.fn(),
    beforePopState: jest.fn(),
    prefetch: jest.fn(),
    push: jest.fn(),
    reload: jest.fn(),
    replace: jest.fn(),
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
    isLocaleDomain: false,
    isReady: false,
    defaultLocale: 'en',
    domainLocales: [],
    isPreview: false,
    ...router,
  };
}
