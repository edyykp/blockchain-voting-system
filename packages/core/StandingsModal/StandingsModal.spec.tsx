import { cleanup, render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { RouterContext } from 'next/dist/shared/lib/router-context';
import { createMockRouter } from 'test-utils/createMockRouter';

import { StandingsModal, StandingsModalProps } from './StandingsModal';

type ImageProps = {
  src: string;
  alt: string;
};

const MOCK_PROPS_NO_SHOW = {
  show: false,
  raceName: 'Monaco',
  circuitId: 'monaco',
  setShowModal: () => {},
};

const MOCK_PROPS_SHOW = {
  show: true,
  raceName: 'Monaco',
  circuitId: 'monaco',
  setShowModal: () => {},
};

const MOCK_FETCHED_DATA = {
  error: null,
  drivers: [
    {
      startingPosition: 18,
      finalPosition: 16,
      Driver: {
        permanentNumber: '6',
        driverId: 'latifi',
        givenName: 'Nicholas',
        familyName: 'Latifi',
        nationality: 'Canadian',
      },
      Constructor: {
        constructorId: 'williams',
        name: 'Williams',
      },
    },
    {
      startingPosition: 10,
      finalPosition: 17,
      Driver: {
        permanentNumber: '14',
        driverId: 'alonso',
        givenName: 'Fernando',
        familyName: 'Alonso',
        nationality: 'Spanish',
      },
      Constructor: {
        constructorId: 'alpine',
        name: 'Alpine F1 Team',
      },
    },
    {
      startingPosition: 2,
      finalPosition: null,
      Driver: {
        permanentNumber: '33',
        driverId: 'max_verstappen',
        givenName: 'Max',
        familyName: 'Verstappen',
        nationality: 'Dutch',
      },
      Constructor: {
        constructorId: 'red_bull',
        name: 'Red Bull',
      },
    },
    {
      startingPosition: 17,
      finalPosition: null,
      Driver: {
        permanentNumber: '5',
        driverId: 'vettel',
        givenName: 'Sebastian',
        familyName: 'Vettel',
        nationality: 'German',
      },
      Constructor: {
        constructorId: 'aston_martin',
        name: 'Aston Martin',
      },
    },
    {
      startingPosition: 9,
      finalPosition: null,
      Driver: {
        permanentNumber: '55',
        driverId: 'sainz',
        givenName: 'Carlos',
        familyName: 'Sainz',
        nationality: 'Spanish',
      },
      Constructor: {
        constructorId: 'ferrari',
        name: 'Ferrari',
      },
    },
  ],
};

jest.mock(
  'next/image',
  () =>
    function Image({ src, alt }: ImageProps) {
      return <img src={src} alt={alt} />;
    },
);

const setup = (props: StandingsModalProps) => {
  return {
    ...render(
      <RouterContext.Provider value={createMockRouter({})}>
        <StandingsModal {...props} />
      </RouterContext.Provider>,
    ),
  };
};

describe('StandingsModalâ†’', () => {
  beforeEach(
    (global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve(MOCK_FETCHED_DATA),
      }),
    ) as jest.Mock),
  );

  afterEach(cleanup);

  it('renders', async () => {
    const { container, getByTestId } = setup(MOCK_PROPS_SHOW);

    await waitFor(() => {
      expect(getByTestId('standings-container')).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });

  it('renders nothing if show is false', async () => {
    const { container, queryByTestId } = setup(MOCK_PROPS_NO_SHOW);

    await waitFor(() => {});
    expect(queryByTestId('standings-container')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders one row for each driver received', async () => {
    const { queryAllByTestId } = setup(MOCK_PROPS_SHOW);

    await waitFor(() => {
      expect(queryAllByTestId('driver-row')).toHaveLength(5);
    });
  });

  it('renders nothing if fetching data failed', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => ({
          drivers: [],
          error: 'This is an error',
        }),
      }),
    ) as jest.Mock;
    const { queryByTestId } = setup(MOCK_PROPS_NO_SHOW);

    await waitFor(() => {});

    expect(queryByTestId('standings-container')).not.toBeInTheDocument();
  });
});
