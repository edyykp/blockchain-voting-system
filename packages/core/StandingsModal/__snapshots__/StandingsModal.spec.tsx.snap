// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandingsModal→ renders 1`] = `
<div>
  <div
    class="overlay"
  >
    <div
      class="wrapper"
    >
      <div
        class="header"
      >
        <h1
          class="title"
        >
          Standings for Monaco
        </h1>
        <svg
          class="icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 352 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          />
        </svg>
      </div>
      <div
        class="body"
      >
        <div
          class="container"
          data-testid="standings-container"
        >
          <div
            class="driverWrapper "
            data-testid="driver-row"
          >
            <div
              class="positionWrapper"
            >
              1
              .
            </div>
            <div
              class="nameWrapper"
            >
              <span
                class="permanentNumber"
              >
                6
              </span>
               
              Nicholas
               
              Latifi
            </div>
            <div
              class="nationalityWrapper"
            >
              Canadian
            </div>
            <div
              class="teamWrapper"
            >
              <span
                class="imageWrapper"
              >
                <img
                  alt="williams"
                  src="/test-file-stub"
                />
              </span>
              <span
                class="constructorWrapper"
              >
                Williams
              </span>
            </div>
          </div>
          <div
            class="driverWrapper greyBackground"
            data-testid="driver-row"
          >
            <div
              class="positionWrapper"
            >
              2
              .
            </div>
            <div
              class="nameWrapper"
            >
              <span
                class="permanentNumber"
              >
                14
              </span>
               
              Fernando
               
              Alonso
            </div>
            <div
              class="nationalityWrapper"
            >
              Spanish
            </div>
            <div
              class="teamWrapper"
            >
              <span
                class="imageWrapper"
              >
                <img
                  alt="alpine"
                  src="/test-file-stub"
                />
              </span>
              <span
                class="constructorWrapper"
              >
                Alpine F1 Team
              </span>
            </div>
          </div>
          <div
            class="driverWrapper "
            data-testid="driver-row"
          >
            <div
              class="positionWrapper"
            >
              3
              .
            </div>
            <div
              class="nameWrapper"
            >
              <span
                class="permanentNumber"
              >
                33
              </span>
               
              Max
               
              Verstappen
            </div>
            <div
              class="nationalityWrapper"
            >
              Dutch
            </div>
            <div
              class="teamWrapper"
            >
              <span
                class="imageWrapper"
              >
                <img
                  alt="red_bull"
                  src="/test-file-stub"
                />
              </span>
              <span
                class="constructorWrapper"
              >
                Red Bull
              </span>
            </div>
          </div>
          <div
            class="driverWrapper greyBackground"
            data-testid="driver-row"
          >
            <div
              class="positionWrapper"
            >
              4
              .
            </div>
            <div
              class="nameWrapper"
            >
              <span
                class="permanentNumber"
              >
                5
              </span>
               
              Sebastian
               
              Vettel
            </div>
            <div
              class="nationalityWrapper"
            >
              German
            </div>
            <div
              class="teamWrapper"
            >
              <span
                class="imageWrapper"
              >
                <img
                  alt="aston_martin"
                  src="/test-file-stub"
                />
              </span>
              <span
                class="constructorWrapper"
              >
                Aston Martin
              </span>
            </div>
          </div>
          <div
            class="driverWrapper "
            data-testid="driver-row"
          >
            <div
              class="positionWrapper"
            >
              5
              .
            </div>
            <div
              class="nameWrapper"
            >
              <span
                class="permanentNumber"
              >
                55
              </span>
               
              Carlos
               
              Sainz
            </div>
            <div
              class="nationalityWrapper"
            >
              Spanish
            </div>
            <div
              class="teamWrapper"
            >
              <span
                class="imageWrapper"
              >
                <img
                  alt="ferrari"
                  src="/test-file-stub"
                />
              </span>
              <span
                class="constructorWrapper"
              >
                Ferrari
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`StandingsModal→ renders nothing if show is false 1`] = `<div />`;
